app/controllers/mixins/topology_mixin.rb

```diff
@@ -0,0 +1,29 @@
+module TopologyMixin
+  def show
+    # When navigated here without id, it means this is a general view for all providers (not for a specific provider)
+    # all previous navigation should not be displayed in breadcrumbs as the user could arrive from
+    # any other page in the application.
+    @breadcrumbs.clear if params[:id].nil?
+    drop_breadcrumb(:name => _('Topology'), :url => '')
+    @lastaction = 'show'
+    @display = @showtype = 'topology'
+  end
+
+  def index
+    redirect_to :action => 'show'
+  end
+
+  def data
+    render :json => {:data => generate_topology(params[:id])}
+  end
+
+  private
+
+  def set_session_data
+    session[:layout] = @layout
+  end
+
+  def generate_topology(provider_id)
+    ContainerTopologyService.new(provider_id).build_topology
```

Comment:
generic mixin contains a call to a specific container provider topology service, that doesn't seem right